sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -0,0 +1,216 @@
+ * `fromType`, respectively. If this is true then it means we can safely cast `value` to `fromType`
+ * and thus able to move the cast to the literal side.
+ *
+ * If the `value` is not within range `(min, max)`, the rule breaks the scenario into different
+ * cases and try to replace each with simpler constructs.
+ *
+ * if `value > max`, the cases are of following:
+ *  - `cast(fromExp, toType) > value` ==> if(isnull(fromExp), null, false)
+ *  - `cast(fromExp, toType) >= value` ==> if(isnull(fromExp), null, false)
+ *  - `cast(fromExp, toType) === value` ==> if(isnull(fromExp), null, false)
+ *  - `cast(fromExp, toType) <=> value` ==> false
+ *  - `cast(fromExp, toType) <= value` ==> if(isnull(fromExp), null, true)
+ *  - `cast(fromExp, toType) < value` ==> if(isnull(fromExp), null, true)
+ *
+ * if `value == max`, the cases are of following:
+ *  - `cast(fromExp, toType) > value` ==> if(isnull(fromExp), null, false)
+ *  - `cast(fromExp, toType) >= value` ==> fromExp == max
+ *  - `cast(fromExp, toType) === value` ==> fromExp == max
+ *  - `cast(fromExp, toType) <=> value` ==> fromExp == max
+ *  - `cast(fromExp, toType) <= value` ==> if(isnull(fromExp), null, true)
+ *  - `cast(fromExp, toType) < value` ==> fromExp =!= max
+ *
+ * Similarly for the cases when `value == min` and `value < min`.
+ *
+ * Further, the above `if(isnull(fromExp), null, false)` is represented using conjunction
+ * `and(isnull(fromExp), null)`, to enable further optimization and filter pushdown to data sources.
+ * Similarly, `if(isnull(fromExp), null, true)` is represented with `or(isnotnull(fromExp), null)`.
+ */
+object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case l: LogicalPlan =>
+      l transformExpressionsUp {
+        case e @ BinaryComparison(_, _) => unwrapCast(e)
+      }
+  }
+
+  private def unwrapCast(exp: Expression): Expression = exp match {
+    case BinaryComparison(Literal(_, _), Cast(_, _, _)) =>
+      // Not a canonical form. In this case we first canonicalize the expression by swapping the
+      // literal and cast side, then process the result and swap the literal and cast again to
+      // restore the original order.
+      def swap(e: Expression): Expression = e match {
+        case GreaterThan(left, right) => LessThan(right, left)
+        case GreaterThanOrEqual(left, right) => LessThanOrEqual(right, left)
+        case EqualTo(left, right) => EqualTo(right, left)
+        case EqualNullSafe(left, right) => EqualNullSafe(right, left)
+        case LessThanOrEqual(left, right) => GreaterThanOrEqual(right, left)
+        case LessThan(left, right) => GreaterThan(right, left)
+        case _ => e
+      }
+
+      swap(unwrapCast(swap(exp)))
+
+    case BinaryComparison(Cast(fromExp, _: IntegralType, _), Literal(value, toType: IntegralType))
+        if canImplicitlyCast(fromExp, toType) =>
```

Comment:
Are we going to add different logics for different types in `canImplicitlyCast`? It might be hard to maintain. We already ensure `toType` is `IntegralType`, can we just check if `fromExp.dataType is instance of IntegralType`. Thus, we can get rid of `canImplicitlyCast`.

Reply:
I do plan to add more checks in this method, for instance checking whether a bigint can be cast to float/double based on the digits. I'm fine to remove it for now and add it back later.

Reply:
maybe we can check in `simplifyIntegral`. If doesn't pass the check, just return `exp`.

Reply:
You mean move `canImplicitlyCast` to `simplifyIntegral` or inline the method in `simplifyIntegral`?

Yeah we could, although later on we may still want to extract these checks into a separate function to improve readability. We can either get rid of `canImplicitlyCast` right now and just replace it with inline checks in `simplifyIntegral`, or keep it and add more checks to it later. I'm fine with either way since this code piece is evolving and will be changed later.